File: JavaGolSource/org/baf/gol/Game.java
package org.baf.gol;

import java.awt.GridLayout;
import java.awt.Rectangle;
import java.awt.image.BufferedImage;
import java.io.BufferedOutputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Date;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import javax.imageio.ImageIO;
import javax.imageio.stream.MemoryCacheImageOutputStream;
import javax.swing.ImageIcon;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JScrollPane;
import javax.swing.border.TitledBorder;

import org.baf.gol.Server.ResponseDataSender;

/**
 * Represents a GoL Game with a set of Game runs.
 */
public class Game {
  public static Game coreGame; // global instance

  static int threadId;

  private int nextThreadId() {
    return ++threadId;
  }

  // play history
  public Map<String, GameRun> runs = new LinkedHashMap<>();

  public int maxCycles = 25; // max that can be played
  public int threadCount; // thread to use in timings
  ParameterProvider pp; // source of command line parameters
  public String saveImageRoot = "/temp"; // change per OS type

  public Game(ParameterProvider pp) {
    this.pp = pp;
  }

  /**
   * Represents a single run of a GoL Game.
   */
  public class GameRun {
    static final int offIndex = 255, onIndex = 0;
    static final int midvalue = 256 / 2; // separates black vs. white

    public Game parent;
    public String name;
    public String imageUrl;
    public Date startedAt, endedAt;
    public int width, height;
    public Grid initialGrid, currentGrid, finalGrid;
    public List<GameCycle> cycles = new ArrayList<>();
    public int delayIn10ms, playIndex;
    public int threadCount;

    private String author = "Unknown";

    public String getAuthor() {
      return author;
    }

    public void setAuthor(String author) {
      this.author = author;
    }

    public GameRun(Game game, String name, String url) throws Exception {
      this.parent = game;
      this.name = name;
      this.imageUrl = url;
      this.delayIn10ms = 5 * 100;

      // make the game grid and load initial state
      String[] kind = new String[1];
      BufferedImage img = Utility.loadImage(url, kind);
      Logger.log.tracef("Image kind: %s", kind[0]);
      if (!"png".equals(kind[0].toLowerCase())) {
        throw new IllegalArgumentException(String.format("named image %s is not a PNG", url));
      }
      var bounds = new Rectangle(img.getMinX(), img.getMinY(), img.getWidth(), img.getHeight());
      var size = bounds.getSize();
      initialGrid = new Grid(size.width, size.height);
      width = initialGrid.width;
      height = initialGrid.height;
      initGridFromImage(bounds.x, bounds.y, bounds.width, bounds.height, img);
      currentGrid = initialGrid.deepClone();
    }

    @Override
    public String toString() {
      return "GameRun[name=" + name + ", imageUrl=" + imageUrl + ", startedSt=" + startedAt + ", endedAt=" + endedAt
          + ", width=" + width + ", height=" + height + ", cycles=" + cycles + ", delayIn10ms=" + delayIn10ms
          + ", playIndex=" + playIndex + ", threadCount=" + threadCount + "]";
    }

    private void initGridFromImage(int minX, int minY, int maxX, int maxY, BufferedImage img) {
      for (int y = minY; y < maxY; y++) {
        for (int x = minX; x < maxX; x++) {
          var pixel = img.getRGB(x, y);
          int r = (pixel >> 16) & 0xFF, g = (pixel >> 8) & 0xFF, b = (pixel >> 0) & 0xFF;

          var cv = 0; // assume all dead
          if (r + g + b < midvalue * 3) {
            cv = 1; // make cell alive
          }
          initialGrid.setCell(x, y, cv);
        }
      }
    }

    public void sendPng(ResponseDataSender rs, int index, int mag) throws IOException {
      Grid grid = null;
      switch (index) {
        case 0:
          grid = initialGrid;
          break;
        default:
          index--;
          if (index < 0 || index >= cycles.size()) {
            throw new ArrayIndexOutOfBoundsException("bad index");
          }
          grid = cycles.get(index).afterGrid;
      }

      var img = new BufferedImage(width * mag + 1, height * mag + 1, BufferedImage.TYPE_BYTE_BINARY);
      fillImage(grid, mag, img);
      var b = encodePngImage(img);
      rs.sendResponseData(b);
      showImageInGui(img); // show in GUI

      if (parent.pp.saveImageFlag()) {
        var saveFile = String.format(saveImageRoot + "/Image_%s.gif", name);
        Files.write(Paths.get(saveFile), b);
        Logger.log.tracef("Save %s", saveFile);
      }
    }

    private byte[] encodePngImage(BufferedImage img) throws IOException {
      var baos = new ByteArrayOutputStream();
      var bos = new BufferedOutputStream(baos);
      var ios = new MemoryCacheImageOutputStream(bos);
      ImageIO.write(img, "png", ios);
      ios.flush();
      return baos.toByteArray();
    }

    private void fillImage(Grid grid, int mag, BufferedImage img) {
      for (var row = 0; row < grid.height; row++) {
        for (var col = 0; col < grid.width; col++) {
          var index = grid.getCell(col, row) != 0 ? onIndex : offIndex;
          // apply magnification
          for (var i = 0; i < mag; i++) {
            for (var j = 0; j < mag; j++) {
              img.setRGB(mag * col + i, mag * row + j, index == onIndex ? 0 : 0x00FFFFFF);
            }
          }
        }
      }
    }

    /**
     * Run a game.
     */
    public void run() {
      this.threadCount = coreGame.threadCount;
      startedAt = new Date();
      int maxCycles = parent.maxCycles;
      for (int count = 0; count < maxCycles; count++) {
        nextCycle();
      }
      endedAt = new Date();
      Logger.log.tracef("GameRun total time: %dms, cycles: %d, thread count: %d",
          endedAt.getTime() - startedAt.getTime(), maxCycles, threadCount);
      finalGrid = currentGrid.deepClone();
    }

    // Advance and play next game cycle.
    // Updating of cycle grid rows can be done in parallel;
    // which can reduce execution time.
    private void nextCycle() {
      var gc = new GameCycle(this);
      gc.beforeGrid = currentGrid.deepClone();
      var p = gc.parent;
      var threadCount = Math.max(p.parent.threadCount, 1);
      gc.afterGrid = new Grid(gc.beforeGrid.width, gc.beforeGrid.height);
      gc.startedAt = new Date();
      var threads = new ArrayList<Thread>();
      var rowCount = (height + threadCount / 2) / threadCount;
      for (var i = 0; i < threadCount; i++) {
        var xi = i;
        var t = new Thread(() -> {
          procesRows(gc, rowCount, xi * rowCount, gc.beforeGrid, gc.afterGrid);
        }, "thread_" + nextThreadId());
        threads.add(t);
        t.setDaemon(true);
        t.start();
      }
      for (var t : threads) {
        try {
          t.join();
        } catch (InterruptedException e) {
          // ignore
        }
      }
      gc.endedAt = new Date();
      currentGrid = gc.afterGrid.deepClone();
      cycles.add(gc);
      gc.cycleCount = cycles.size();
    }

    // process all cells in a set of rows
    private void procesRows(GameCycle gc, int rowCount, int startRow, Grid inGrid, Grid outGrid) {
      for (var index = 0; index < rowCount; index++) {
        var rowIndex = index + startRow;
        for (var colIndex = 0; colIndex < width; colIndex++) {
          // count any neighbors
          var neighbors = 0;
          if (inGrid.getCell(colIndex - 1, rowIndex - 1) != 0) {
            neighbors++;
          }
          if (inGrid.getCell(colIndex, rowIndex - 1) != 0) {
            neighbors++;
          }
          if (inGrid.getCell(colIndex + 1, rowIndex - 1) != 0) {
            neighbors++;
          }
          if (inGrid.getCell(colIndex - 1, rowIndex) != 0) {
            neighbors++;
          }
          if (inGrid.getCell(colIndex + 1, rowIndex) != 0) {
            neighbors++;
          }
          if (inGrid.getCell(colIndex - 1, rowIndex + 1) != 0) {
            neighbors++;
          }
          if (inGrid.getCell(colIndex, rowIndex + 1) != 0) {
            neighbors++;
          }
          if (inGrid.getCell(colIndex + 1, rowIndex + 1) != 0) {
            neighbors++;
          }
          // determine next generation cell state based on neighbor count
          var pv = inGrid.getCell(colIndex, rowIndex);
          var nv = 0;
          switch (neighbors) {
            case 2:
              nv = pv;
              break;
            case 3:
              if (pv == 0) {
                nv = 1;
              }
              break;
          }
          outGrid.setCell(colIndex, rowIndex, nv);
        }
      }
    }

    /**
     * Make images from 1+ cycles into GIF form.
     */
    public byte[] makeGifs(int count, int mag) throws IOException {
      var cycleCount = cycles.size();
      var xcycles = Math.min(count, cycleCount + 1);
      List<BufferedImage> bia = new ArrayList<>();
      var added = addGridSafe(initialGrid, 0, xcycles, mag, bia);
      for (int i = 0; i < cycleCount; i++) {
        added = addGridSafe(cycles.get(i).afterGrid, added, xcycles, mag, bia);
      }
      return packGifs(added, mag, delayIn10ms, bia.toArray(new BufferedImage[bia.size()]));
    }

    int addGridSafe(Grid grid, int added, int max, int mag, List<BufferedImage> bia) {
      var img = new BufferedImage(mag * width + 1, mag * height + 1, BufferedImage.TYPE_BYTE_BINARY);
      if (added < max) {
        fillImage(grid, mag, img);
        bia.add(img);
        added++;
      }
      return added;
    }

    byte[] packGifs(int count, int mag, int delay, BufferedImage[] bia) throws IOException {
      showImagesInGui(bia);

      var baos = new ByteArrayOutputStream();
      var bos = new BufferedOutputStream(baos);
      var ios = new MemoryCacheImageOutputStream(bos);
      AnnimatedGifWriter.createGifs(ios, delay, author, bia);
      ios.flush();
      return baos.toByteArray();
    }

    // not in Go version.
    void showImagesInGui(BufferedImage[] bia) {
      // create a Swing Frame to show a row of images
      var frame = new JFrame("Show Images rendered at " + new Date());
      frame.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
      JPanel imagePanel = new JPanel(new GridLayout());
      var sp = new JScrollPane(imagePanel);
      frame.setContentPane(sp);
      frame.setSize(1000, 800);

      var index = 1;
      for (var bi : bia) {
        var icon = new ImageIcon(bi);
        JLabel labelledIcon = new JLabel(icon);
        labelledIcon.setBorder(
            new TitledBorder(String.format("Image: %d (%dx%d)", index++, icon.getIconWidth(), icon.getIconHeight())));
        imagePanel.add(labelledIcon);
      }
      frame.setVisible(true);
    }

    // not in Go version.
    void showImageInGui(BufferedImage bi) {
      var frame = new JFrame("Show Image rendered at " + new Date());
      JPanel imagePanel = new JPanel(new GridLayout());
      var sp = new JScrollPane(imagePanel);
      frame.setContentPane(sp);
      frame.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
      frame.setSize(1000, 800);
      var icon = new ImageIcon(bi);
      JLabel labelledIcon = new JLabel(icon);
      labelledIcon
          .setBorder(new TitledBorder(String.format("Image: (%dx%d)", icon.getIconWidth(), icon.getIconHeight())));
      imagePanel.add(labelledIcon);
      frame.setVisible(true);
    }
  }

  /**
   * Clear all runs.
   */
  public void clear() {
    runs.clear();
  }

  /**
   * Run a game.
   */
  public void run(String name, String url) throws Exception {
    var gr = new GameRun(this, name, url);
    runs.put(gr.name, gr);
    gr.run();
  }

  /**
   * Represents a GoL Game grid.
   */
  public static class Grid {
    public byte[] data;
    public int width, height;

    public Grid(int width, int height) {
      this.width = width;
      this.height = height;
      data = new byte[width * height];
    }

    @Override
    public String toString() {
      return "Grid[width=" + width + ", height=" + height + "]";
    }

    public int getCell(int x, int y) {
      if (x < 0 || x >= width || y < 0 || y >= height) {
        return 0;
      }
      return data[x + y * width];
    }

    public void setCell(int x, int y, int cv) {
      if (x < 0 || x >= width || y < 0 || y >= height) {
        return;
      }
      data[x + y * width] = (byte) cv;
    }

    public Grid deepClone() {
      var ng = new Grid(width, height);
      for (int i = 0; i < data.length; i++) {
        ng.data[i] = data[i];
      }
      ng.width = width;
      ng.height = height;
      return ng;
    }
  }

  /**
   * Represents a GoL Game cycle.
   */
  public static class GameCycle {
    public GameRun parent;
    public int cycleCount;
    public Date startedAt, endedAt;
    public Grid beforeGrid, afterGrid;

    public GameCycle(GameRun parent) {
      this.parent = parent;
    }

    @Override
    public String toString() {
      return "GameCycle[cycle=" + cycleCount + ", " + "startedAt=" + startedAt + ", endedAt=" + endedAt + "]";
    }
  }

}

File: GoGolSource/main/game.go
package main

import (
	"bytes"
	"errors"
	"fmt"
	"image"
	"image/color"
	"image/gif"
	"image/png"
	"io"
	"io/ioutil"
	"log"
	"os"
	"sync"
	"time"
)

// Default game history.
var CoreGame = &Game{
	make(map[string]*GameRun),
	10,
	0,
	1}

// Represents a game.
type Game struct {
	Runs           map[string]*GameRun
	MaxCycles      int
	SkipCycles     int // not currently used
	GoroutineCount int
}

// Run a set of cycles from the grid defined by an image.
func (g *Game) Run(name, url string) (err error) {
	gr, err := NewGameRun(name, url, g)
	if err != nil {
		return
	}
	g.Runs[gr.Name] = gr
	err = gr.Run()
	return
}

// Clear a game.
func (g *Game) Clear() {
	for k, _ := range g.Runs {
		delete(g.Runs, k)
	}
}

// Represents a single run of a game.
type GameRun struct {
	Parent         *Game
	Name           string
	ImageURL       string
	StartedAt      time.Time
	EndedAt        time.Time
	Width, Height  int
	InitialGrid    *Grid
	CurrentGrid    *Grid
	FinalGrid      *Grid
	Cycles         []*GameCycle
	DelayIn10ms    int
	PlayIndex      int
	GoroutineCount int
}

// B & W color indexes
const (
	offIndex = 0
	onIndex  = 1
)

// B & W color palette
var paletteBW = []color.Color{color.White, color.Black}

// Generate a PNG result (single frame).
func (gr *GameRun) MakePNG(writer io.Writer, index int) (err error) {
	var grid *Grid
	switch index {
	case 0:
		grid = gr.InitialGrid
	default:
		index--
		if index < 0 || index >= len(gr.Cycles) {
			err = BadIndexError
			return
		}
		grid = gr.Cycles[index].AfterGrid
	}
	mag := magFactorFlag
	rect := image.Rect(0, 0, mag*gr.Width+1, mag*gr.Height+1)
	img := image.NewPaletted(rect, paletteBW)
	gr.FillImage(grid, img)
	b, err := gr.encodePNGImage(img)
	if err != nil {
		return
	}
	count, err := writer.Write(b.Bytes())
	log.Printf("Returned PNG, size= %d\n", count)
	if saveImageFlag {
		saveFile := fmt.Sprintf("/temp/Image_%s_%d.png", gr.Name, index)
		xerr := ioutil.WriteFile(saveFile, b.Bytes(), os.ModePerm)
		fmt.Printf("Save %s: %v\n", saveFile, xerr)
	}
	return
}

// Make a PNG image.
func (gr *GameRun) encodePNGImage(img *image.Paletted) (b bytes.Buffer, err error) {
	var e png.Encoder
	e.CompressionLevel = png.NoCompression
	err = e.Encode(&b, img)
	return
}

// Generate a GIF result (>= 1 frame).
func (gr *GameRun) MakeGIFs(count int) (agif *gif.GIF, err error) {
	mag := magFactorFlag
	cycles := len(gr.Cycles)
	xcount := cycles + 1
	if xcount > count {
		xcount = count
	}
	added := 0
	agif = &gif.GIF{LoopCount: 5}

	rect := image.Rect(0, 0, mag*gr.Width+1, mag*gr.Height+1)
	img := image.NewPaletted(rect, paletteBW)
	if added < xcount {
		gr.AddGrid(gr.InitialGrid, img, agif)
		added++
	}
	for i := 0; i < cycles; i++ {
		if added < xcount {
			img = image.NewPaletted(rect, paletteBW)
			gc := gr.Cycles[i]
			grid := gc.AfterGrid
			gr.AddGrid(grid, img, agif)
			added++
		}
	}
	return
}

// Fill in and record a cycle image in an animated GIF.
func (gr *GameRun) AddGrid(grid *Grid, img *image.Paletted, agif *gif.GIF) {
	gr.FillImage(grid, img)
	agif.Image = append(agif.Image, img)
	agif.Delay = append(agif.Delay, gr.DelayIn10ms)
}

// Fill in an image from a grid.
func (gr *GameRun) FillImage(grid *Grid, img *image.Paletted) {
	mag := magFactorFlag
	for row := 0; row < grid.Height; row++ {
		for col := 0; col < grid.Width; col++ {
			index := offIndex
			if grid.getCell(col, row) != 0 {
				index = onIndex
			}
			// apply magnification
			for i := 0; i < mag; i++ {
				for j := 0; j < mag; j++ {
					img.SetColorIndex(mag*row+i, mag*col+j, uint8(index))
				}
			}
		}
	}
}

const midValue = 256 / 2 // middle color value

// Error values.
var (
	NotPNGError   = errors.New("not a png")
	NotRGBAError  = errors.New("not RGBA color")
	BadIndexError = errors.New("bad index")
)

// Start a new game run.
func NewGameRun(name, url string, parent *Game) (gr *GameRun, err error) {
	gr = &GameRun{}
	gr.Parent = parent
	gr.Name = name
	gr.GoroutineCount = CoreGame.GoroutineCount
	gr.ImageURL = url
	gr.DelayIn10ms = 5 * 100
	var img image.Image
	var kind string
	img, kind, err = LoadImage(url)
	if err != nil {
		return
	}
	fmt.Printf("Image kind:  %v\n", kind)
	if kind != "png" {
		return nil, NotPNGError
	}
	bounds := img.Bounds()
	minX, minY, maxX, maxY := bounds.Min.X, bounds.Min.Y, bounds.Max.X, bounds.Max.Y
	size := bounds.Size()
	//xsize := size.X * size.Y
	gr.InitialGrid = NewEmptyGrid(size.X, size.Y)
	gr.Width = gr.InitialGrid.Width
	gr.Height = gr.InitialGrid.Height

	err = gr.InitGridFromImage(minX, maxX, minY, maxY, img)
	if err != nil {
		return
	}
	gr.CurrentGrid = gr.InitialGrid.DeepCloneGrid()
	return
}

// Fill in a grid from an image.
// Map color images to B&W.  Only RGBA images allowed.
func (gr *GameRun) InitGridFromImage(minX, maxX, minY, maxY int,
	img image.Image) (err error) {
	setCount, totalCount := 0, 0
	for y := minY; y < maxY; y++ {
		for x := minX; x < maxX; x++ {
			//			r, g, b, a := img.At(x, y).RGBA()
			rgba := img.At(x, y)
			var r, g, b uint8
			switch v := rgba.(type) {
			case color.NRGBA:
				r, g, b, _ = v.R, v.G, v.B, v.A
			case color.RGBA:
				r, g, b, _ = v.R, v.G, v.B, v.A
			default:
				err = NotRGBAError
				return
			}
			cv := byte(0) // assume cell dead
			if int(r)+int(g)+int(b) < midValue*3 {
				cv = byte(1) // make cell alive
				setCount++
			}
			gr.InitialGrid.setCell(x, y, cv)
			totalCount++
		}
	}
	return
}

// Play a game.
// Run requested cycle count.
func (gr *GameRun) Run() (err error) {
	gr.StartedAt = time.Now()
	for count := 0; count < gr.Parent.MaxCycles; count++ {
		err = gr.NextCycle()
		if err != nil {
			return
		}
	}
	gr.EndedAt = time.Now()
	fmt.Printf("GameRun total time: %dms, goroutine count: %d\n",
		(gr.EndedAt.Sub(gr.StartedAt)+NanosPerMs)/NanosPerMs, gr.GoroutineCount)
	gr.FinalGrid = gr.CurrentGrid.DeepCloneGrid()
	return
}

// Represents a single cycle of a game.
type GameCycle struct {
	Parent     *GameRun
	Cycle      int
	StartedAt  time.Time
	EndedAt    time.Time
	BeforeGrid *Grid
	AfterGrid  *Grid
}

func NewGameCycle(parent *GameRun) (gc *GameCycle) {
	gc = &GameCycle{}
	gc.Parent = parent
	return
}

// Advance and play next game cycle.
// Updating of cycle grid rows can be done in parallel;
// which can reduce execution time.
func (gr *GameRun) NextCycle() (err error) {
	gc := NewGameCycle(gr)
	gc.BeforeGrid = gr.CurrentGrid.DeepCloneGrid()
	p := gc.Parent
	goroutineCount := p.Parent.GoroutineCount
	if goroutineCount <= 0 {
		goroutineCount = 1
	}
	gc.AfterGrid = NewEmptyGrid(gc.BeforeGrid.Width, gc.BeforeGrid.Height)
	gc.StartedAt = time.Now()
	// process rows across  allowed goroutines
	rowCount := (gr.Height + goroutineCount/2) / goroutineCount
	var wg sync.WaitGroup
	for i := 0; i < goroutineCount; i++ {
		wg.Add(1)
		go processRows(&wg, gc, rowCount, i*rowCount, gc.BeforeGrid, gc.AfterGrid)
	}
	wg.Wait() // let all finish
	gc.EndedAt = time.Now()
	gr.CurrentGrid = gc.AfterGrid.DeepCloneGrid()
	gr.Cycles = append(gr.Cycles, gc)
	gc.Cycle = len(gr.Cycles)
	return
}

// Represents a 2-dimensional game grid (abstract, not as an image).
type Grid struct {
	Data          []byte
	Width, Height int
}

func NewEmptyGrid(w, h int) (g *Grid) {
	g = &Grid{}
	g.Data = make([]byte, w*h)
	g.Width = w
	g.Height = h
	return
}

func (g *Grid) DeepCloneGrid() (c *Grid) {
	c = &Grid{}
	lg := len(g.Data)
	c.Data = make([]byte, lg, lg)
	for i, b := range g.Data {
		c.Data[i] = b
	}
	c.Width = g.Width
	c.Height = g.Height
	return
}

func (g *Grid) getCell(x, y int) (b byte) {
	if x < 0 || x >= g.Width || y < 0 || y >= g.Height {
		return
	}
	return g.Data[x+y*g.Width]
}
func (g *Grid) setCell(x, y int, b byte) {
	if x < 0 || x >= g.Width || y < 0 || y >= g.Height {
		return
	}
	g.Data[x+y*g.Width] = b
}

// Play game as subset of grid rows (so can be done in parallel).
func processRows(wg *sync.WaitGroup, gc *GameCycle, rowCount int,
	startRow int, inGrid, outGrid *Grid) {
	defer wg.Done()
	gr := gc.Parent
	for index := 0; index < rowCount; index++ {
		rowIndex := index + startRow
		for colIndex := 0; colIndex < gr.Width; colIndex++ {
			// count any neighbors
			neighbors := 0
			if inGrid.getCell(colIndex-1, rowIndex-1) != 0 {
				neighbors++
			}
			if inGrid.getCell(colIndex, rowIndex-1) != 0 {
				neighbors++
			}
			if inGrid.getCell(colIndex+1, rowIndex-1) != 0 {
				neighbors++
			}
			if inGrid.getCell(colIndex-1, rowIndex) != 0 {
				neighbors++
			}
			if inGrid.getCell(colIndex+1, rowIndex) != 0 {
				neighbors++
			}
			if inGrid.getCell(colIndex-1, rowIndex+1) != 0 {
				neighbors++
			}
			if inGrid.getCell(colIndex, rowIndex+1) != 0 {
				neighbors++
			}
			if inGrid.getCell(colIndex+1, rowIndex+1) != 0 {
				neighbors++
			}

			// determine next generation cell state based on neighbor count
			pv := inGrid.getCell(colIndex, rowIndex)
			nv := uint8(0) // assume dead
			switch neighbors {
			case 2:
				nv = pv // unchanged
			case 3:
				if pv == 0 {
					nv = 1 // make alive
				}
			}
			outGrid.setCell(colIndex, rowIndex, nv)
		}
	}
}
